labrador/src/prover.rs

```diff
@@ -1,85 +1,1210 @@
-// src/prover.rs
+use profiler_macro::time_profiler;
+use rand::Rng;
+use crate::setup::setup;
+use crate::algebra::{PolynomialRing, RqMatrix, Zq};
 
-pub fn setup() {
-    // 0. setup
-    // public parameters after setup: [a_ij^(k), a_ij^(l), phi^(k), phi^(l), b^(k), b0(l)']
 
-    // 1. setup constraints
-    // 1.1 get s_i and do norm check
-    // 1.1.1 get s_i = s_1 - s_r; r is the number of witness s
-    // each s_i is a vector of ring elements
+// inner product of 2 vectors of PolynomialRing
+fn inner_product_polynomial_ring_vector(
+    a: &Vec<PolynomialRing>,
+    b: &Vec<PolynomialRing>,
+) -> PolynomialRing {
+    a.iter()
+        .zip(b.iter())
+        .map(|(a, b)| a * b)
+        .collect::<Vec<PolynomialRing>>()
+        .into_iter()
+        .reduce(|acc, x| acc + x)
+        .unwrap()
+}
+
+fn inner_product_zq_vector(a: &Vec<Zq>, b: &Vec<Zq>) -> Zq {
+    a.iter()
+        .zip(b.iter())
+        .map(|(a, b)| *a * *b)
+        .sum()
+}
 
-    // 1.1.2 get beta norm bound, refer paper page 26, theorem 6.3
```

Comment:
Some suggestions: 
1. Perhaps using Full Constraint and Constant Constraint to replace b_constraint and b'_constraint would make it clearer.
2. Perhaps using Dot Product to replace Inner Product would align better with the notation in the paper.

We could try adding comments before each function to annotate the parameters it uses. Since there are quite a few parameters in lattice-related implementations, this approach should make the code more readable.
Ex.
Notation:
- b_constraint: b^(k)
- b'_constraint: b'^(l)
- s = witness vector
- a_constraint = a_ij^(k): quadratic
- phi_constraint = phi^(k): linear
- b_consraint = b^(k): constant

// Function to calculate b^(k)

Reply:
update: discussed with @pinhaocypher we can keep current convention